### PR TITLE
Refactor get analytical jacobian

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54480 Implement faster gradcheck and enable for all ops except some
* #54479 Factor out numerical logic
* #54378 Refactor get numerical jacobian to calculate wrt all outputs at once
* #54092 Refactor get numerical jacobian
* **#54049 Refactor get analytical jacobian**

### For release notes:
`torch.autograd.gradcheck.get_analytical_jacobian` (not part of the public api) is being deprecated.

--- end --

The goal of this is to factor out the core logic of getting the analytical jacobian which is effectively doing `f(grad_out) = grad_out^T J = grad_input`. This allows us to test a lot of logic that was not possible before because now we can replace f with whatever we want in order to simulate potential issues that gradcheck is designed to catch.

Edit: I realize a lot of things this PR was originally aiming to allow is actually possible with hooks, hence the tests have already been added in a earlier PR in the stack. But this is still slightly useful for reducing code duplication when adding the new fast gradcheck code (more details below)

After this change, `get_analytical_jacobian` is only responsible for gathering a list of rows that are later combined into a single Jacobian tensor. This means we don't have to perform any checks for correctness of the dtypes/size at this step

We factor out that logic into a separate function, `combine_jacobian_rows`, which handles the list of rows -> single Tensor step for each jacobian, and the error checking it entails. (This allows this code to be shared between the fast/slow versions.)

Differential Revision: [D27307240](https://our.internmc.facebook.com/intern/diff/D27307240)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -161,13 +161,54 @@ def get_stride(size):
...
+def check_analytical_jacobian_attributes(inputs, output, nondet_tol, grad_out_scale, check_grad_dtypes,
+                                         raise_exception, custom_backward_fn=None):
+    diff_input_list = list(iter_tensors(inputs, True))
+
+    def backward_fn(grad_output):
+        return torch.autograd.grad(output, diff_input_list, grad_output,
+                                   retain_graph=True, allow_unused=True)
+    fn = custom_backward_fn if custom_backward_fn is not None else backward_fn
+    jacobians_rows = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
+    jacobians_rows_reentrant = get_analytical_jacobian(fn, output.clone(), grad_out_scale)
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:203]: I never really understood what the reentrant check is testing. What is it testing?

The reason why I am asking is because previously, we would compute row 0 of the first jacobian, then row 0 of the second jacobian, then row 1 of the first jacobian, then row 1 of the second jacobian, etc. This PR changes it so that we compute the full first jacobian followed by the full second jacobian. Does this affect the reentrant check?

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
...
-            reentrant = False
-
-    return jacobian, reentrant, correct_grad_sizes, correct_grad_types
+    return jacobians, failed
+
+
+def get_analytical_jacobian(fn, sample_output, grad_out_scale):
+    # Computes Jacobian row-by-row using backward function `fn` = v^T J
+    # NB: we can't combine the rows into a single jacobian tensor because fn(v) for
+    # different v may return tensors with different number of elements
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:233]: Is this because we expect fn(v) to return tensors with the same number of elements if the gradient formula is correct but are not making that assumption because it can be wrong?

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
...
-
-    for jacobian_x, jacobian_reentrant_x in zip(jacobian, jacobian_reentrant):
-        if jacobian_x.numel() != 0 and (jacobian_x - jacobian_reentrant_x).abs().max() > nondet_tol:
-            reentrant = False
-
-    return jacobian, reentrant, correct_grad_sizes, correct_grad_types
+    return jacobians, failed
+
+
+def get_analytical_jacobian(fn, sample_output, grad_out_scale):
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:230]: nit: fn -> vjp_fn is more descriptive (v^T J is a vector-jacobian-product)

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -183,43 +224,25 @@ def fail_test(msg):
...
+    # Computes Jacobian row-by-row using backward function `fn` = v^T J
+    # NB: we can't combine the rows into a single jacobian tensor because fn(v) for
+    # different v may return tensors with different number of elements
+    grad_out_base = torch.zeros_like(sample_output, memory_format=torch.legacy_contiguous_format)
+    flat_grad_out = grad_out_base.view(-1)
+    # jacobians_rows[i][j] represents the jth row of the ith input
+    jacobians_rows: Dict[int, List[Optional[torch.Tensor]]] = {}
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:237]: Why does it need to be a Dict[int, List[Optional[torch.Tensor]]]? Couldn't it be a List[List[Optional[torch.Tensor]]]?

File: torch/autograd/gradcheck.py [hunk 5]
```diff
@@ -161,13 +161,54 @@ def get_stride(size):
...
+            if row is not None and row.size() != inp.size():
+                correct_grad_sizes = False
+            elif row is not None and row.dtype != inp.dtype:
+                correct_grad_types = False
+            if row is None:
+                out_jacobian[:, j].zero_()
+            else:
+                row_dense = row.to_dense() if not row.layout == torch.strided else row
+                assert out_jacobian[:, j].numel() == row_dense.numel()
+                out_jacobian[:, j] = row_dense.contiguous().view(-1)
```
Comment (zou3519) [inline torch/autograd/gradcheck.py:190]: nit: row_dense.reshape(-1) does the same thing

**Review by zou3519 — COMMENTED** [review]
refactor looks correct, I had a few questions

File: torch/autograd/gradcheck.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:203]: I don't really know what its named 'reentrant' check, AFAICT its just computing something twice and checking if they are the same as to see if its deterministic. A better name imo should be determinism check not reentrant check. Maybe @albanD knows more.

File: torch/autograd/gradcheck.py (hunk 4 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:237]: Yeah, its probably safe to assume that different grad_outs to backward produce the same number of inputs.

File: torch/autograd/gradcheck.py (hunk 1 above)
Comment (zou3519) [reply torch/autograd/gradcheck.py]: Offline @albanD said this is re-entrant in the sense that it is calling the same function twice. This is different from when we call autograd from within autograd, which we also refer to as "re-entrant". So yes, looks like a case of bad naming

Comment (zou3519) [reply torch/autograd/gradcheck.py]: Also I realized that the code does explain this in an error message: https://github.com/pytorch/pytorch/blob/87989a6cf9979d3842991ab1696a62a1d247a523/torch/autograd/gradcheck.py#L472-L475

**Review by zou3519 — APPROVED** [review]
